File: boot-start/src/main/java/com/example/demo/DemoApplication.java
package com.example.demo;


import java.net.URI;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.ToString;
import lombok.extern.slf4j.Slf4j;
import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Component;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

@SpringBootApplication
public class DemoApplication {

    public static void main(String[] args) {
        SpringApplication.run(DemoApplication.class, args);
    }
}

@RestController
@RequestMapping(value = "/posts")
class PostController {

    private final PostRepository posts;

    public PostController(PostRepository posts) {
        this.posts = posts;
    }

    @GetMapping(value = "")
    public Flux<Post> all() {
        return this.posts.findAll();
    }

    @GetMapping(value = "/{id}")
    public Mono<Post> get(@PathVariable(value = "id") UUID id) {
        return this.posts.findById(id);
    }

    @PostMapping(value = "")
    public Mono<ResponseEntity<?>> create(@RequestBody Post post) {
        return this.posts.save(post).map(p -> ResponseEntity.created(URI.create("/posts/" + p.getId())).build());
    }

}

@Component
@Slf4j
class PostRepository {

    private static final List<Post> DATA = new ArrayList<>();

    static {
        DATA.add(Post.builder().id(UUID.randomUUID()).title("post one").content("content of post one").build());
        DATA.add(Post.builder().id(UUID.randomUUID()).title("post two").content("content of post two").build());
    }

    Flux<Post> findAll() {
        return Flux.fromIterable(DATA);
    }


    Mono<Post> findById(UUID id) {
        return findAll().filter(p -> p.getId().equals(id)).last();
    }

    Mono<Post> save(Post post) {
        Post saved = Post.builder().id(UUID.randomUUID()).title(post.getTitle()).content(post.getContent()).build();
        log.debug("saved post: {}", saved);
        DATA.add(saved);
        return Mono.just(saved);
    }

}

@Data
@ToString
@Builder
@NoArgsConstructor
@AllArgsConstructor
class Post {
    private UUID id;
    private String title;
    private String content;
}

File: boot-kotlin-co/src/main/kotlin/com/example/demo/DemoApplication.kt
package com.example.demo

import kotlinx.coroutines.flow.Flow
import org.slf4j.LoggerFactory
import org.springframework.boot.SpringApplication
import org.springframework.boot.autoconfigure.SpringBootApplication
import org.springframework.context.event.ContextRefreshedEvent
import org.springframework.context.event.EventListener
import org.springframework.data.annotation.Id
import org.springframework.data.mongodb.core.mapping.Document
import org.springframework.data.repository.kotlin.CoroutineCrudRepository
import org.springframework.http.HttpStatus
import org.springframework.http.ProblemDetail
import org.springframework.http.ResponseEntity
import org.springframework.stereotype.Component
import org.springframework.web.bind.annotation.*
import java.net.URI
import java.time.LocalDateTime

@SpringBootApplication
class DemoApplication

fun main(args: Array<String>) {
    SpringApplication.run(DemoApplication::class.java, *args)
}

@RestController
@RequestMapping(value = ["/posts"])
class PostController(val posts: PostRepository) {

    @GetMapping("")
    fun all(): ResponseEntity<Flow<Post>> = ResponseEntity.ok(this.posts.findAll())

    @PostMapping("")
    suspend fun create(@RequestBody post: Post): ResponseEntity<Void> {
        val saved = this.posts.save(post)
        return ResponseEntity.created(URI.create("/posts/${saved.id}")).build()
    }

    @GetMapping("/{id}")
    suspend fun get(@PathVariable("id") id: String): ResponseEntity<Post> {
        val existed = this.posts.findById(id) ?: throw PostNotFoundException(id)
        return ResponseEntity.ok(existed)
    }

    @PutMapping("/{id}")
    suspend fun update(@PathVariable("id") id: String, @RequestBody post: Post): ResponseEntity<Void> {
        val existed = this.posts.findById(id) ?: throw PostNotFoundException(id)
        existed.apply {
            title = post.title
            content = post.content
        }

        this.posts.save(existed)
        return ResponseEntity.noContent().build()
    }

    @DeleteMapping("/{id}")
    suspend fun delete(@PathVariable("id") id: String): ResponseEntity<Void> {
        this.posts.deleteById(id)
        return ResponseEntity.noContent().build()
    }

    @ExceptionHandler(value = [PostNotFoundException::class])
    suspend fun handlePostNotFoundException(e: PostNotFoundException): ProblemDetail {
        val detail = ProblemDetail.forStatusAndDetail(HttpStatus.NOT_FOUND, e.message ?: "Post was not found")
        detail.apply {
            title = "Not Found"
            type = URI.create("http://example.com/not_found")
        }
        return detail
    }

}

class PostNotFoundException(id: String) : RuntimeException("Post $id was not found")

@Component
class DataInitializer(val posts: PostRepository) {
    companion object {
        private val log = LoggerFactory.getLogger(DataInitializer::class.java)
    }

    @EventListener(value = [ContextRefreshedEvent::class])
    suspend fun run() {
        log.info("start data initialization ...")

        val samplePosts = listOf("Post one", "Post two").map { Post(title = it, content = "content of $it") }
        posts.saveAll(samplePosts).collect { log.debug("saved post: $it") }

        log.info("data initialization is done ...")
    }
}

interface PostRepository : CoroutineCrudRepository<Post, String>

@Document
data class Post(
    @Id var id: String? = null,
    var title: String? = null,
    var content: String? = null,
    var createdDate: LocalDateTime = LocalDateTime.now()
)